src/gui/qgsmapcanvas.cpp

```diff
@@ -604,8 +604,10 @@ void QgsMapCanvas::stopRendering()
   {
     QgsDebugMsg( "CANVAS stop rendering!" );
     mJobCanceled = true;
-    mJob->cancel();
-    Q_ASSERT( !mJob ); // no need to delete here: already deleted in finished()
+    disconnect( mJob, &QgsMapRendererJob::finished, this, &QgsMapCanvas::rendererJobFinished );
+    connect( mJob, &QgsMapRendererQImageJob::finished, mJob, &QgsMapRendererQImageJob::deleteLater );
+    mJob->cancelWithoutBlocking();
+    mJob = nullptr;
```

Comment:
Does this crash on exit if the canvas is destroyed while we are still waiting for a job to be cancelled?

I tried to do this in the past with a slightly different approach (which I abandoned because I had rendering artefacts from multiple dying jobs painting to the same cache image in parallel).

This particular part of it could be interesting anyway to keep track of zombie threads: https://github.com/opengisch/QField/blob/master/src/qgsquickmapcanvasmap.cpp#L400